### PR TITLE
Support for vSMP MemoryOne in conjunction with SuSE JeOS/CHost

### LINKED ISSUE #18: Support cloud-config for the MemoryOne OS.
**What would you like to be added**: 
MemoryOne OS allows using cost-effective non-volatile storage-class memory (SCM) and other high-performance storage devices to transparently replace expensive server DRAM and expand beyond their capacity limits. Read more [here](https://www.scalemp.com/products/vsmp-memoryone/) 

In order to support MemoryOne based machines with the gardener, jeos/chost extension needs to generate following cloud-config for `chost-memoryone`[name to be decided] operating system. 

```
Content-Type: multipart/mixed; boundary="==BOUNDARY=="
MIME-Version: 1.0
--==BOUNDARY==
Content-Type: text/x-vsmp; section=vsmp
system_memory=6x
mem_topology=2
--==BOUNDARY==
Content-Type: text/x-shellscript
#!/bin/bash
​
zypper addrepo https://download.opensuse.org/repositories/Archiving/SLE_15_SP1/Archiving.repo
zypper --non-interactive --no-gpg-checks install bzip2
​
mkdir /run/vsmp
wget "http://www.scalemp.com/downloads/others/EC2/latest/vsmp_installer.sh" -O /run/vsmp/vsmp_installer.sh
chmod 755 /run/vsmp/vsmp_installer.sh
/run/vsmp/vsmp_installer.sh in -tq
​
<STANDARD_CHOST_USERDATA_HERE>
​
--==BOUNDARY==
```

It would be nice to have the `system_memory` and `mem_topology` parameters configurable via shoot-resource unless it adds unnecessary complexity.

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area os
/kind enhancement
/priority normal
/os suse-chost

**What this PR does / why we need it**:
This PR adds support for the [vSMP MemoryOne](https://www.scalemp.com/products/vsmp-memoryone/) operating system in conjunction with SuSE JeOS/CHost as described in #18.

**Which issue(s) this PR fixes**:
Fixes #18 
Ref https://github.com/gardener/gardener/issues/2354

**Special notes for your reviewer**:
* ✅ Depends on https://github.com/gardener/gardener/pull/2420
* ✅ Depends on https://github.com/gardener/gardener/pull/2429
* In a follow-up PR we should finally rename the extension to `suse-chost` as JeOS/CHost is inconsistently used across the whole project.

/assign @vpnachev @hardikdr 
/invite @vpnachev 

**Release note**:

```noteworthy user
This extension controller is now supporting the [vSMP MemoryOne](https://www.scalemp.com/products/vsmp-memoryone/) operating system in conjunction with SuSE JeOS/CHost. You can find more information in [this document](https://github.com/gardener/gardener-extension-os-suse-jeos/blob/master/docs/usage-as-end-user.md#support-for-vsmp-memoryone).
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
-	sigs.k8s.io/controller-runtime v0.4.0
+	k8s.io/apimachinery v0.17.6
+	k8s.io/code-generator v0.17.6
+	k8s.io/utils v0.0.0-20200327001022-6496210b90e8
+	sigs.k8s.io/controller-runtime v0.5.4
 )
 
 replace (
+	github.com/gardener/gardener => github.com/rfranzke/gardener v0.0.0-20200604105041-85655d065cb5
 	k8s.io/api => k8s.io/api v0.16.8 // 1.16.8
```
Comment (ialidzhikov) [inline go.mod:26]: Can you update also to k/k `v0.17.6` under the replace directive?

**Comment by ialidzhikov** [comment]
I think the reason for the ci failure is that currently `.ci/verify` miss GOPATH workaround such as https://github.com/gardener/gardener-extension-provider-aws/blob/v1.8.2/.ci/verify#L12-L15. (as this PR introduces code-generation for this repo)

File: docs/usage-as-end-user.md [hunk 2]
```diff
@@ -1,19 +1,108 @@
...
+    - eu-central-1b
+```
+
+### Generating an AWS snapshot ID for the JeOS/CHost operating system
+
+The following script will help to generate the snapshot ID on AWS:
+
+```bash
+# Script to automate the generation of the snapshot-id.
+# TODO not needed if AMI is provided
```
Comment (vpnachev) [inline docs/usage-as-end-user.md:88]: I do not understand why it is necessary and what is expected to be improved?

File: docs/usage-as-end-user.md [hunk 3]
```diff
@@ -1,19 +1,108 @@
...
+    zones:
+    - eu-central-1b
+```
+
+### Generating an AWS snapshot ID for the JeOS/CHost operating system
+
+The following script will help to generate the snapshot ID on AWS:
+
+```bash
+# Script to automate the generation of the snapshot-id.
```
Comment (vpnachev) [inline docs/usage-as-end-user.md:87]: All resources here are scoped to the default region defined in $HOME/.aws/config.
Please, add a note that the script must be run for every region where `memoryone-chost` nodes are needed.

File: docs/usage-as-end-user.md [hunk 4]
```diff
@@ -1,19 +1,108 @@
...
+  fi
+  echo "AMI $name ... creating private snapshot"
+  inst=`aws ec2 run-instances --instance-type t3.nano --image-id $ami --query 'Instances[0].InstanceId' --output=text`
+  aws ec2 wait instance-running --instance-ids $inst
+  vol=`aws ec2 describe-instances --instance-ids $inst --query "Reservations[].Instances[].BlockDeviceMappings[0].Ebs.VolumeId" --output=text`
+  snap=`aws ec2 create-snapshot --description "snap-$name" --volume-id $vol --query='SnapshotId' --tag-specifications "ResourceType=snapshot,Tags=[{Key=Name,Value=\"$name\"}]" --output=text`
+  aws ec2 wait snapshot-completed --snapshot-ids $snap
+  aws ec2 terminate-instances --instance-id $inst > /dev/null
+  # root=`aws ec2 describe-instances --instance-ids $inst --query "Reservations[].Instances[].BlockDeviceMappings[0].DeviceName" --output=text`
+  #aws ec2 register-image --name "copy-$name" --description "copy-$name" --block-device-mapping "[{\"DeviceName\": \"$root\",\"Ebs\":{\"SnapshotId\":\"$snap\",\"DeleteOnTermination\": true}}]" --root-device-name $root --output=text
```
Comment (vpnachev) [inline docs/usage-as-end-user.md:105]: Please, remove commented code that is about to be executed by end-users.

File: pkg/apis/memoryonechost/v1alpha1/types.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// OperatingSystemConfiguration allows to specify configuration for the operating system.
+type OperatingSystemConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// MemTopology allows to configure the `mem_topology` parameter. If not present, it will default to `2`.
+	// +optional
+	MemTopology *string `json:"memTopology,omitempty"`
```
Comment (vpnachev) [inline pkg/apis/memoryonechost/v1alpha1/types.go:30]: How about ?

```suggestion
	MemoryTopology *string `json:"memoryTopology,omitempty"`
```

File: pkg/generator/cloudinit_test.go [hunk 6]
```diff
@@ -12,11 +12,20 @@
...
 	"os"
 
+	oscommongenerator "github.com/gardener/gardener/extensions/pkg/controller/operatingsystemconfig/oscommon/generator"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+
+	"github.com/gardener/gardener-extension-os-suse-jeos/pkg/apis/memoryonechost"
+	"github.com/gardener/gardener-extension-os-suse-jeos/pkg/generator"
+	"github.com/gardener/gardener-extension-os-suse-jeos/pkg/susejeos"
```
Comment (vpnachev) [inline pkg/generator/cloudinit_test.go:27]: I think this packages should be in the second import block.

File: docs/usage-as-end-user.md (hunk 2 above)
Comment (rfranzke) [reply docs/usage-as-end-user.md]: left-over from copied script, I didn't check it in detail and simply added it to the docs. fixed now, thanks

**Pull request merged**